conn.go

```diff
@@ -481,13 +491,17 @@ func (hc *halfConn) encrypt(b *block, explicitIVLen int) (bool, alert) {
 				hc.additionalData[11] = byte(payloadLen >> 8)
 				hc.additionalData[12] = byte(payloadLen)
 				additionalData = hc.additionalData[:]
-			}
-
-			if hc.version >= VersionTLS13 {
+			} else {
 				// opaque type
 				payload = payload[:len(payload)+1]
 				payload[len(payload)-1] = b.data[0]
 				b.data[0] = byte(recordTypeApplicationData)
+
+				// Add AD header, see 5.2 of RFC8446
+				additionalData = make([]byte, 5)
+				additionalData[0] = byte(recordTypeApplicationData)
+				binary.BigEndian.PutUint16(additionalData[1:], VersionTLS12)
+				binary.BigEndian.PutUint16(additionalData[3:], uint16(payloadLen+overhead))
```

Comment:
I wouldn't overload the word `overhead` here. This is supposed to be the tag length of the underlying AEAD, i.e., the output of `c.Overhead()`. I'd define a variable `recordLen`. If `hc.version >= VersionTLS13`, then let `recordLen = payloadLen + 1`. Then on  this line, replace `uint16(payloadLen+overhead))` with `uint16(recordLen+c.Overhead())`.

Reply:
I think there is more refactoring than that needed for this entire function. this is coming in next PR.  This one I would like to keep minimal.

Reply:
Sounds good.